### PR TITLE
Closes #4988: Load URL if user chooses not to open in external app

### LINKED ISSUE #4988: Use SessionUseCases in AppLinksFeature loadUrl
Currently AppLinksFeature loads URL by dirrectly calling `loadUrl`.  However, it should be using SessionUseCases `loadUrl`.  Update AppLinksFeature to loadUrl using SessionUseCases.

### PR DESCRIPTION
**Pull request by rocketsroger**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/app-links/src/main/java/mozilla/components/feature/app/links/RedirectDialogFragment.kt [hunk 1]
```diff
@@ -22,6 +22,13 @@ abstract class RedirectDialogFragment : DialogFragment() {
      */
     var onConfirmRedirect: () -> Unit = {}
 
+    /**
+     * A callback to trigger a URL load, call it when you are ready load the URL in the current session.
```
Comment (csadilek) [inline components/feature/app-links/src/main/java/mozilla/components/feature/app/links/RedirectDialogFragment.kt:26]: nit: I would rephrase this. The callback isn't specifically to trigger a load, it's invoked when the user cancels / dismisses the dialog. re: "to trigger a URL load when you are...".

File: components/feature/app-links/src/test/java/mozilla/components/feature/app/links/AppLinksFeatureTest.kt [hunk 2]
```diff
@@ -154,4 +155,13 @@ class AppLinksFeatureTest {
         verify(mockDialog).show(eq(mockFragmentManager), anyString())
         verify(mockOpenRedirect, never()).invoke(any(), anyBoolean(), any())
     }
+
+    @Test
+    fun `reused redirect dialog if exists`() {
```
Comment (csadilek) [inline components/feature/app-links/src/test/java/mozilla/components/feature/app/links/AppLinksFeatureTest.kt:160]: Hmm this doesn't test the logic that's being added here i.e. it doesn't test the new callback?

File: components/feature/app-links/src/main/java/mozilla/components/feature/app/links/RedirectDialogFragment.kt (hunk 1 above)
Comment (rocketsroger) [reply components/feature/app-links/src/main/java/mozilla/components/feature/app/links/RedirectDialogFragment.kt:26]: Sure, I'll rename.

File: components/feature/app-links/src/test/java/mozilla/components/feature/app/links/AppLinksFeatureTest.kt (hunk 2 above)
Comment (rocketsroger) [reply components/feature/app-links/src/test/java/mozilla/components/feature/app/links/AppLinksFeatureTest.kt:160]: I'll test for the new callback.  Thanks,

File: components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksFeature.kt [hunk 3]
```diff
@@ -23,14 +25,17 @@ import mozilla.components.support.base.feature.LifecycleAwareFeature
...
- * @param sessionId the session ID to observe.
+ * @param sessionId The session ID to observe.
  * @param fragmentManager FragmentManager for interacting with fragments.
  * @param dialog The dialog for redirect.
  * @param launchInApp If {true} then launch app links in third party app(s). Default to false because
  * of security concerns.
  * @param useCases These use cases allow for the detection of, and opening of links that other apps
  * have registered to open.
+ * @param failedToLaunchAction Action to perform when failing to launch in third party app
+ * @param loadUrlUseCase Used to load URL if user decides not to launch in third party app
```
Comment (csadilek) [inline components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksFeature.kt:36]: nit++ period. :)

File: components/feature/app-links/src/test/java/mozilla/components/feature/app/links/SimpleRedirectDialogFragmentTest.kt [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        dialog.show()
+
+        val confirmButton = dialog.findViewById<Button>(android.R.id.button2)
+        confirmButton?.performClick()
+        assertFalse(onConfirmCalled)
+        assertTrue(onCancelCalled)
+    }
+
+    @Test
+    fun `Dialog confirm and cancel is not called when dismissed`() {
```
Comment (csadilek) [inline components/feature/app-links/src/test/java/mozilla/components/feature/app/links/SimpleRedirectDialogFragmentTest.kt:71]: Just curious, what is the behaviour here if the user dismisses? We won't be launching the app, and also won't be loading the URL, right? So essentially, we do nothing in that case?

File: components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksFeature.kt (hunk 3 above)
Comment (rocketsroger) [reply components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksFeature.kt:36]: Will fix, thanks

File: components/feature/app-links/src/test/java/mozilla/components/feature/app/links/SimpleRedirectDialogFragmentTest.kt (hunk 4 above)
Comment (rocketsroger) [reply components/feature/app-links/src/test/java/mozilla/components/feature/app/links/SimpleRedirectDialogFragmentTest.kt:71]: Yes, though we `setCancelable` to false, but the user can still back out or tap outside the dialog to dismiss it.  Do you want us to load the URL if the user dismisses the dialog without tapping on the buttons?

Comment (rocketsroger) [reply components/feature/app-links/src/test/java/mozilla/components/feature/app/links/SimpleRedirectDialogFragmentTest.kt:71]: Figured out that the fragment should be set to false in `setCancelable`.  This prevents user tapping on the outside of the dialog to dismiss it.  In practice the user shouldn't be able to dismiss the dialog anymore.  I'll keep the test though since its on the dialog level.

Comment (csadilek) [reply components/feature/app-links/src/test/java/mozilla/components/feature/app/links/SimpleRedirectDialogFragmentTest.kt:71]: No, I am fine :). Was just curious about it.

**Comment by rocketsroger** [comment]
bors r=csadilek
